src/hypothesis/searchstrategy/datetime.py

```diff
@@ -0,0 +1,96 @@
+# coding=utf-8
+#
+# This file is part of Hypothesis, which may be found at
+# https://github.com/HypothesisWorks/hypothesis-python
+#
+# Most of this work is copyright (C) 2013-2017 David R. MacIver
+# (david@drmaciver.com), but it contains contributions by others. See
+# CONTRIBUTING.rst for a full list of people who may hold copyright, and
+# consult the git log if you need to determine who owns an individual
+# contribution.
+#
+# This Source Code Form is subject to the terms of the Mozilla Public License,
+# v. 2.0. If a copy of the MPL was not distributed with this file, You can
+# obtain one at http://mozilla.org/MPL/2.0/.
+#
+# END HEADER
+
+from __future__ import division, print_function, absolute_import
+
+import datetime as dt
+
+from hypothesis.internal.conjecture import utils
+from hypothesis.searchstrategy.strategies import SearchStrategy
+
+__all__ = ['DatetimeStrategy', 'TimedeltaStrategy']
+
+
+class DatetimeStrategy(SearchStrategy):
+
+    def __init__(self, min_datetime, max_datetime, timezones_strat):
+        assert isinstance(min_datetime, dt.datetime)
+        assert isinstance(max_datetime, dt.datetime)
+        assert min_datetime.tzinfo is None
+        assert max_datetime.tzinfo is None
+        assert min_datetime <= max_datetime
+        assert isinstance(timezones_strat, SearchStrategy)
+        self.min_dt = min_datetime
+        self.max_dt = max_datetime
+        self.tz_strat = timezones_strat
+
+    def do_draw(self, data):
+        # This looks like it could loop forever, but in the worst case it
+        # would fail after exhausting the underlying entropy.  Obviously this
+        # is still bad, but even pathological bounds only loop a few times.
+        while True:
+            result = dict()
+            low_bound = True
+            high_bound = True
+            for name in ('year', 'month', 'day',
+                         'hour', 'minute', 'second', 'microsecond'):
+                low = getattr(
+                    self.min_dt if low_bound else dt.datetime.min, name)
+                high = getattr(
+                    self.max_dt if high_bound else dt.datetime.max, name)
+                if name == 'year':
+                    val = utils.centered_integer_range(data, low, high, 2000)
+                else:
+                    val = utils.integer_range(data, low, high)
+                result[name] = val
+                low_bound = low_bound and val == low
+                high_bound = high_bound and val == high
+            tz = data.draw(self.tz_strat)
+            try:
+                result = dt.datetime(**result)
+                if hasattr(tz, 'localize') and hasattr(tz, 'normalize'):
+                    # Can't just construct; see http://pytz.sourceforge.net
+                    return tz.normalize(tz.localize(result))
+                # TODO: do dateutil or other libraries have similar issues?
```

Comment:
Having dug through the docs, it looks like stdlib and dateutil timezone objects do not need special handling.